### PR TITLE
[LIBCLOUD-879] Add support of node without public IP in LB

### PR DESCRIPTION
**Pull request by charly37**
### Description

Link to jira ticket 879 
https://issues.apache.org/jira/browse/LIBCLOUD-879

I had a condition to only try to grab the publicIP of a node if it had one since publicIP is not mandatory.

### Status

This PR is ready to be review. I run the test (tox) before and after the change and no impact.

### Checklist (tick everything that applies)

- [X] [Code linting] I tried to followed the guidline
- [X] [Documentation] No change require
- [X] [Tests] Added 1 test case to cover the new behavior
- [X] [ICLA] Minor change so I skip it

### CODE REVIEW

File: libcloud/loadbalancer/drivers/gce.py [hunk 1]
```diff
@@ -338,9 +338,12 @@ def _node_to_member(self, node, balancer):
         # would be found if it was there.
         if hasattr(node, 'name'):
             member_id = node.name
-            member_ip = node.public_ips[0]
         else:
             member_id = node
+
+        if (hasattr(node, 'public_ips') and (len(node.public_ips) > 0)):
```
Comment (tonybaloney) [inline libcloud/loadbalancer/drivers/gce.py]: no need for parens here and it's not very Pythonic to use them either ;-)

**Comment by charly37** [comment]
Updated following @tonybaloney review.

**Comment by charly37** [comment]
I was wandering if I need to provide any extra information to move forward on this PR following the changes I made last week ?

File: libcloud/loadbalancer/drivers/gce.py [hunk 2]
```diff
@@ -338,9 +338,12 @@ def _node_to_member(self, node, balancer):
         # would be found if it was there.
         if hasattr(node, 'name'):
             member_id = node.name
-            member_ip = node.public_ips[0]
         else:
             member_id = node
+
```
Comment (allardhoeve) [inline libcloud/loadbalancer/drivers/gce.py:343]: Although a correct solution, I think we should solve this by fixing the `Node` object, which should always have a `public_ips` attribute that is empty if there are no public IPs. This has the benefit of us needing to code that only once instead of all over the place.

Also, I would like to see a testcase, so this functionality never gets accidentally removed.

Comment (allardhoeve) [reply libcloud/loadbalancer/drivers/gce.py:343]: Testcases are added by adding the Google API output for a node without public_ips to the fixture, then adding a test to the Node object to make sure the node has an empty list, then this function needs no conditionals, just:

```python
try:
    member_ip = node.public_ips[0]
except IndexError:
    member_ip = None
```

**Review by allardhoeve — CHANGES_REQUESTED** [review]

File: libcloud/loadbalancer/drivers/gce.py (hunk 2 above)
Comment (allardhoeve) [reply libcloud/loadbalancer/drivers/gce.py:343]: The reason I ask for a testcase is that it makes it possible to refactor this kind of thing without introducing new bugs.

**Review by allardhoeve — CHANGES_REQUESTED** [review]
I'd like a new test for this behaviour and also some refactoring of the logic into the `Node` object.

**Comment by allardhoeve** [comment]
Hey Charles, sorry to spring more work on you after 20 days, but I'd like this to keep working next year :)

File: libcloud/loadbalancer/drivers/gce.py (hunk 2 above)
Comment (charly37) [reply libcloud/loadbalancer/drivers/gce.py:343]: @allardhoeve Thx a lot for the feedback. 

1/ node object : In fact the node object is good.  The node IP has indeed an empty "publicIP" which lead to the error since we try to access the first element. Maybe I did not fully understand your feedback ?
2/ test case : I can definitively try to add a test case (will be a good experience for me). 

Will update the PR with a test case ASAP and wait for your clarification about the point 1/

Thx again guys for the feedback.

Comment (allardhoeve) [reply libcloud/loadbalancer/drivers/gce.py:343]: 1/ you check if the `Node` object has a `public_ips` attribute, why? Does it not always have the attribute? If not, we should make sure it always has the attribute. If so, you can remove the `hasattr`.

Comment (charly37) [reply libcloud/loadbalancer/drivers/gce.py:343]: Thx @allardhoeve for your clarification. 
In fact checking the length of "public_ips" was my first idea but it do not work (break some test case). The "Node" object can in fact be a string and not a proper node object. That is probably why the actual code test for the attribute "name". This is also written in a comment that I saw after my initial test break the regression 
```
        # A balancer can have a node as a member, even if the node doesn't
        # exist.  In this case, 'node' is simply a string to where the resource
        # would be found if it was there.
```
That is why I keep the same design as the actual code where the attribute "name" is tested before accessing it for my new change. I hope it clarify ?
PS : test case in progress
PSS : Node is a string when the load balancer reference a non longer existing node

Comment (allardhoeve) [reply libcloud/loadbalancer/drivers/gce.py:343]: Huh, I think that is not a really good design from a user's perspective. Maybe we can check with the original author what the intent was? I think having incompatible mixed types is a smell, so maybe we should set node to none and have a different field for the original node name?

Comment (charly37) [reply libcloud/loadbalancer/drivers/gce.py:343]: I think it would be indeed possible to design it another way to avoid mixing String and Node objects nevertheless it would change the actual behavior (that could be use by some existing application) and that is why I limited my change to fixing the publicIP issue with the actual design. 

What if I open another dedicated JIRA ticket to discuss this point ?

**Comment by charly37** [comment]
@allardhoeve I added a test case as suggested.

File: libcloud/test/loadbalancer/test_gce.py [hunk 3]
```diff
@@ -188,6 +188,12 @@ def test_node_to_member(self):
         self.assertEqual(member.ip, node.public_ips[0])
         self.assertEqual(member.id, node.name)
         self.assertEqual(member.port, balancer.port)
+        # Test node without publicIP
```
Comment (allardhoeve) [inline libcloud/test/loadbalancer/test_gce.py]: Can you make this a test of its own?

**Review by allardhoeve — CHANGES_REQUESTED** [review]
:+1: looks good, but I think tests should be separate.

**Comment by allardhoeve** [comment]
Thanks Charles! I'll merge this momentarily.

**Comment by allardhoeve** [comment]
The tests are failing, but that has nothing to do with this PR

**Comment by allardhoeve** [comment]
@tonybaloney, any idea why the tests might be failing? They fail on `apt-get install libvirt`, but I cannot find that provisioning code in the repo anywhere.

**Comment by charly37** [comment]
@allardhoeve Tests are failing because of the Travis setup. Looking at some other PR it is the same issue but it seems someone find a workaround by using libvirt-bin package instead of libvirt (see https://github.com/apache/libcloud/commit/20ab4cbf89c855a2e8b67f49ce239de506eab865 )I was thinking to do it in my PR too but i m afraid it generate conflict.

File: libcloud/test/loadbalancer/test_gce.py (hunk 3 above)
Comment (charly37) [reply libcloud/test/loadbalancer/test_gce.py]: Sure. Will do it today

**Comment by allardhoeve** [comment]
Please try to merge master into this branch or rebase against it, thanks.

**Comment by charly37** [comment]
@allardhoeve Tests have been split as suggest. Rebase from trunk to fix the TRAVIS jobs.
I think it is all clean to be merge ;)

**Comment by allardhoeve** [comment]
Yes! Will merge today

**Comment by charly37** [comment]
Thx again for the review and help.
I also open a ticket to discuss the design of mixing string and node object as discuss previously in the PR review : https://issues.apache.org/jira/browse/LIBCLOUD-883

**Comment by allardhoeve** [comment]
Merged, thanks. Should be closed in a minute.

**Comment by pquentin** [comment]
Nice catch @charly37!
